File: junk/string/lcs-naive.cpp
#include <iostream>
#include <string>


using namespace std;

string LCS(string S, string T)
{
  if (S.size() == 0 || T.size() == 0)
    return "";

  if (S[0] == T[0])
    return S[0] + LCS(S.substr(1),  T.substr(1));

  string s1 = LCS(S.substr(1), T);
  string s2 = LCS(S, T.substr(1));

  if (s1.size() > s2.size())
    return s1;

  return s2;
}

int
main(int argc, char *argv[])
{
  cout << "LCS:" << LCS("a1b2c3d4e", "zz1yy2xx3ww4vv") << endl;
  return 0;
}

File: junk/string/lcs-naive.py
#!/usr/bin/python

def LCS(S, T):
    if not len(S) or not len(T):
        return ''

    if S[0] == T[0]:
        return S[0] + LCS(S[1:],  T[1:])

    s1 = LCS(S[1:], T)
    s2 = LCS(S, T[1:])

    if len(s1) > len(s2):
        return s1

    return s2

print LCS('a1b2c3d4e', 'zz1yy2xx3ww4vv')
